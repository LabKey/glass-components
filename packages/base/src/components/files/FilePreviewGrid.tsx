import * as React from 'react';
import {Map, List} from 'immutable';

import { Alert } from "../Alert";
import { Grid, GridColumn } from "../Grid";

interface Props {
    data: List<Map<string, any>>
    columns?: List<GridColumn>
    header?: string
    infoMsg?: any
    errorMsg?: string
}

export class FilePreviewGrid extends React.Component<Props, any> {

    static defaultProps = {
        header: 'File preview:',
        msg:''
    };

    render() {
        const { data, columns, header, infoMsg, errorMsg } = this.props;
        const numRows = data ? data.size : 0;

        return (
            <>
                {errorMsg ? <Alert>{errorMsg}</Alert>
                    : <>
                        <strong>{header}</strong>
                        <p className={'margin-top'}>
                            <span>
                                The {numRows === 1 ? 'one row ' : 'first ' + numRows + ' rows '} of your data
                                file {numRows === 1 ? 'is' : 'are'} shown below.
                            </span>
                            &nbsp;
                            {infoMsg}
                        </p>
                        <Grid columns={columns} data={data}/>
                    </>
                }
            </>
        )
    }
}