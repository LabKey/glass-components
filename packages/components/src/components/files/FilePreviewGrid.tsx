import React from 'react';
import { List, Map } from 'immutable';

import { Alert } from '../base/Alert';
import { Grid, GridColumn } from '../base/Grid';
import { InferDomainResponse } from '../base/models/model';

export interface FileGridPreviewProps {
    previewCount: number
    header?: string
    infoMsg?: any
    onPreviewLoad?: (response: InferDomainResponse, fileData?: File) => any
    acceptedFormats?: string // comma-separated list of allowed extensions i.e. '.png, .jpg, .jpeg'
    initialData?: InferDomainResponse
    skipPreviewGrid?: boolean
}

type Props = FileGridPreviewProps & {
    data: List<Map<string, any>>
    columns?: List<GridColumn>
    errorMsg?: string
}


export class FilePreviewGrid extends React.Component<Props, any> {

    static defaultProps = {
        header: 'File preview:',
        msg:''
    };

    render() {
        const { data, columns, header, infoMsg, errorMsg } = this.props;
        const numRows = data ? data.size : 0;

        return (
            <>
                {errorMsg ? <Alert>{errorMsg}</Alert>
                    : <>
                        <strong>{header}</strong>
                        <p className={'margin-top'}>
                            <span>
                                The {numRows === 1 ? 'one row ' : 'first ' + numRows + ' rows '} of your data
                                file {numRows === 1 ? 'is' : 'are'} shown below.
                            </span>
                            &nbsp;
                            {infoMsg}
                        </p>
                        <Grid columns={columns} data={data}/>
                    </>
                }
            </>
        )
    }
}
