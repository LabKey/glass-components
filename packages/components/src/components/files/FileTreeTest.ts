const data = {
    name: 'root',
    children: [
        {
            name: 'parent1',
            children: [
                {
                    name: 'child1.xlsx',
                    data: { iconFontCls: 'fa fa-file-excel-o' },
                },
                {
                    name: 'child2.pdf',
                    data: { iconFontCls: 'fa fa-file-pdf-o' },
                },
            ],
        },
        {
            name: 'loading parent',
            children: [{ name: 'slow-child1' }, { name: 'slow-child2' }],
        },
        {
            name: 'parent2',
            children: [
                {
                    name: 'nested parent',
                    children: [{ name: 'nested child 1' }, { name: 'nested child 2' }],
                },
            ],
        },
        {
            name: 'empty directory',
            children: [],
        },
    ],
};

const getFileFromPath = (path: string): string => {
    const pathParts = path.split('/');

    return pathParts[pathParts.length - 1];
};

const getDirectoryNode = (name: string, root: any): any => {
    if (root.name === name) {
        return root;
    }

    let node;

    if (root.children) {
        root.children.forEach(child => {
            const found = getDirectoryNode(getFileFromPath(name), child);
            if (found) {
                node = { ...found };
            }
        });
    }

    return node;
};

const wait = ms => new Promise(resolve => window.setTimeout(resolve, ms));

export const fetchFileTestTree = (directory?: string): Promise<any> => {
    return new Promise(resolve => {
        let node = { ...data }; // Make a copy

        if (directory) {
            node = getDirectoryNode(directory, node);

            if (node.children) {
                const newChildren = node.children.map(child => {
                    const newChild = { ...child };
                    if (newChild.children) {
                        newChild.children = []; // empty this out to replicate data coming from webdav
                    }
                    return newChild;
                });

                resolve(
                    node.name === 'loading parent'
                        ? wait(3000).then(() => {
                              return newChildren;
                          })
                        : newChildren
                );
            }
        }

        node.children = [];
        resolve(node);
    });
};
