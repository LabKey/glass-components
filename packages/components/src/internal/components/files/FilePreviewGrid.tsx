import React from 'react';
import { List, Map } from 'immutable';

import { Alert, Grid, GridColumn, InferDomainResponse } from '../../..';
import { FileGridPreviewProps } from '../../../public/files/models';

type Props = FileGridPreviewProps & {
    data: List<Map<string, any>>;
    columns?: List<GridColumn>;
    errorMsg?: string;
};

export class FilePreviewGrid extends React.Component<Props, any> {
    static defaultProps = {
        header: 'File preview:',
        msg: '',
        errorStyle: 'warning',
    };

    render() {
        const { data, columns, header, infoMsg, errorMsg, errorStyle } = this.props;
        const numRows = data ? data.size : 0;

        return (
            <>
                {errorMsg ? (
                    <Alert bsStyle={errorStyle}>{errorMsg}</Alert>
                ) : (
                    <>
                        <strong>{header}</strong>
                        <p className="margin-top">
                            <span>
                                The {numRows === 1 ? 'one row ' : 'first ' + numRows + ' rows '} of your data file{' '}
                                {numRows === 1 ? 'is' : 'are'} shown below.
                            </span>
                            &nbsp;
                            {infoMsg}
                        </p>
                        <Grid columns={columns} data={data} />
                    </>
                )}
            </>
        );
    }
}
