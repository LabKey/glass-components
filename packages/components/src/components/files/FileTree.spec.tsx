import {FilePreviewGrid} from "./FilePreviewGrid";
import {fromJS} from "immutable";
import renderer from "react-test-renderer";
import React from "react";
import {mount, shallow} from "enzyme";
import {AssayDesignerPanels, FileTree} from "../..";
import toJson from "enzyme-to-json";
import {fetchFileTestTree} from "./FileTreeTest";


describe("<FileTree/>", () => {

    const waitForLoad = jest.fn((component) => Promise.resolve(!component.state().loading));

    test("with data", () => {
        const tree = shallow(
            <FileTree loadData={fetchFileTestTree}/>
        );

        return waitForLoad(tree).then(() => {

            let node = tree.childAt(0).dive().childAt(0).dive().find('NodeHeader');
            expect(node.prop('node')['children'].length).toEqual(0);

            node.simulate('click');

            return waitForLoad(tree).then(() => {

                expect(node.prop('node')['children'].length).toEqual(4);
                expect(toJson(tree)).toMatchSnapshot();
                tree.unmount();
            });
        });
    });
});
