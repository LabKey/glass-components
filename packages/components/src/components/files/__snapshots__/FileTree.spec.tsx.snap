// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FileTree/> with data 1`] = `
<div
  className="filetree-container"
>
  <TreeBeard
    animations={
      Object {
        "drawer": [Function],
        "toggle": [Function],
      }
    }
    data={
      Object {
        "active": true,
        "children": Array [
          Object {
            "children": Array [],
            "id": "root|parent1",
            "name": "parent1",
          },
          Object {
            "children": Array [],
            "id": "root|loading parent",
            "loading": true,
            "name": "loading parent",
          },
          Object {
            "children": Array [],
            "id": "root|parent2",
            "name": "parent2",
          },
          Object {
            "children": Array [],
            "id": "root|empty directory",
            "name": "empty directory",
          },
        ],
        "id": "root",
        "name": "root",
        "toggled": true,
      }
    }
    decorators={
      Object {
        "Container": [Function],
        "Header": [Function],
        "Loading": [Function],
        "Toggle": [Function],
      }
    }
    onToggle={[Function]}
    style={
      Object {
        "tree": Object {
          "base": Object {
            "backgroundColor": "white",
            "color": "#777",
            "fontFamily": "lucida grande ,tahoma,verdana,arial,sans-serif",
            "fontSize": "14px",
            "listStyle": "none",
            "margin": 0,
            "padding": 0,
          },
          "node": Object {
            "activeLink": Object {
              "background": "white",
            },
            "base": Object {
              "position": "relative",
            },
            "header": Object {
              "base": Object {
                "color": "#777",
                "display": "inline-block",
                "verticalAlign": "top",
              },
              "connector": Object {
                "borderBottom": "solid 2px black",
                "borderLeft": "solid 2px black",
                "height": "12px",
                "left": "-21px",
                "position": "absolute",
                "top": "0px",
                "width": "2px",
              },
              "title": Object {
                "lineHeight": "24px",
                "verticalAlign": "middle",
              },
            },
            "link": Object {
              "cursor": "pointer",
              "display": "block",
              "padding": "0px 5px",
              "position": "relative",
            },
            "loading": Object {
              "color": "#777",
            },
            "subtree": Object {
              "listStyle": "none",
              "paddingLeft": "19px",
            },
            "toggle": Object {
              "arrow": Object {
                "fill": "#777",
                "strokeWidth": 0,
              },
              "base": Object {
                "display": "inline-block",
                "height": "24px",
                "marginLeft": "-5px",
                "position": "relative",
                "verticalAlign": "top",
                "width": "24px",
              },
              "height": 14,
              "width": 14,
              "wrapper": Object {
                "height": "14px",
                "left": "50%",
                "margin": "-7px 0 0 -7px",
                "position": "absolute",
                "top": "50%",
              },
            },
          },
        },
      }
    }
  />
</div>
`;
